### PR TITLE
Set debug variables from argv

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
Propietary RGSS editors call Game.exe with debug, test or btest arguments (according to issue #190), setting $DEBUG, $TEST or $BTEST respectively as true.

To use it replace Game.exe with the mkxp binary.
Then when using some propietary editor from wine, mkxp will be called instead of a propietary implementation and named flags are included.

### CODE REVIEW

File: binding-mri/binding-mri.cpp [hunk 1]
```diff
@@ -147,7 +147,24 @@ static void mriBindingInit()
...
+	for (int i = 1; i < shState->rtData().argc; i++)
+	{
+		const std::string arg = shState->rtData().argv[i];
+		if ("debug" == arg)
+			has_debug = true;
+		else if ("test" == arg)
+			has_test = true;
+		else if ("btest" == arg)
+			has_btest = true;
+	}
```
Comment (Ancurio) [inline binding-mri/binding-mri.cpp]: The block above is not specific to MRI, it doesn't belong in the binding

File: binding-mri/binding-mri.cpp [hunk 2]
```diff
@@ -147,7 +147,24 @@ static void mriBindingInit()
 	_rb_define_module_function(mod, "raw_key_states", mkxpRawKeyStates);
 	_rb_define_module_function(mod, "mouse_in_window", mkxpMouseInWindow);
 
+	/* Load global constants */
+	bool has_debug = false;
+	bool has_test = false;
+	bool has_btest = false;
```
Comment (Ancurio) [inline binding-mri/binding-mri.cpp]: I don't believe I have used snake_case anywhere in mkxp

File: binding-mri/binding-mri.cpp [hunk 3]
```diff
@@ -147,7 +147,24 @@ static void mriBindingInit()
...
+	for (int i = 1; i < shState->rtData().argc; i++)
+	{
+		const std::string arg = shState->rtData().argv[i];
+		if ("debug" == arg)
```
Comment (Ancurio) [inline binding-mri/binding-mri.cpp]: Please use `variable == "constant"` order; we don't need defensive coding when compilers do a good enough job of catching our errors these days.

File: binding-mri/binding-mri.cpp [hunk 4]
```diff
@@ -147,7 +147,24 @@ static void mriBindingInit()
...
+			has_debug = true;
+		else if ("test" == arg)
+			has_test = true;
+		else if ("btest" == arg)
+			has_btest = true;
+	}
 	rb_gv_set("MKXP", Qtrue);
+	rb_gv_set("DEBUG", has_debug ? Qtrue : Qfalse);
+	rb_gv_set("TEST", has_test ? Qtrue : Qfalse);
+	rb_gv_set("BTEST", has_btest ? Qtrue : Qfalse);
```
Comment (Ancurio) [inline binding-mri/binding-mri.cpp]: There is a convenience function for C++ → Ruby bool conversion in `binding-util.h`

File: binding-mruby/binding-mruby.cpp [hunk 5]
```diff
@@ -114,7 +114,25 @@ static void mrbBindingInit(mrb_state *mrb)
...
+		else if ("test" == arg)
+			has_test = true;
+		else if ("btest" == arg)
+			has_btest = true;
+	}
+
 	mrb_define_global_const(mrb, "MKXP", mrb_true_value());
+	mrb_define_global_const(mrb, "DEBUG", has_debug ? mrb_true_value() : mrb_false_value());
+	mrb_define_global_const(mrb, "TEST", has_test ? mrb_true_value() : mrb_false_value());
+	mrb_define_global_const(mrb, "BTEST", has_btest ? mrb_true_value() : mrb_false_value());
```
Comment (Ancurio) [inline binding-mruby/binding-mruby.cpp]: I think all comments on the MRI binding code apply here as well.

File: src/filesystem.cpp [hunk 6]
```diff
@@ -508,7 +508,7 @@ findFontsFolderCB(void *data, const char *, const char *fname)
 	char buffer[512];
 	const char *s = fname;
 
-	while (s && i < sizeof(buffer))
+	while (*s && i < sizeof(buffer))
```
Comment (Ancurio) [inline src/filesystem.cpp]: Looks like a commit from master sneaked in

**Comment by ReinUsesLisp** [comment]
Sorry for those commits.
Instead of stop boost-parsing when debug, test or btest are found, argv is moved one argument, to allow something like: `$ ./mkxp debug --fullscreen true`
RGSS1 uses $DEBUG and VX uses $TEST ($BTEST is used by both), so not only style was wrong.

I'm writing a free alternative to known privative XP editor, that's why I'm interested in these calls.

**Comment by Ancurio** [comment]
No need to apologize, your commits look very good now! Thanks for working on this. :)  
You should combine all the changes you made so far into one commit, you can do that with `git rebase -i master` and then marking all commits before your newest one as `fixup` or `f`. Then do a `git push --force ...` to update your branch.  
This makes the commit history look clean, and easier for me to review.

File: binding-mruby/binding-mruby.cpp [hunk 7]
```diff
@@ -115,24 +115,9 @@ static void mrbBindingInit(mrb_state *mrb)
...
-			has_test = true;
-		else if ("btest" == arg)
-			has_btest = true;
-	}
-
 	mrb_define_global_const(mrb, "MKXP", mrb_true_value());
-	mrb_define_global_const(mrb, "DEBUG", has_debug ? mrb_true_value() : mrb_false_value());
-	mrb_define_global_const(mrb, "TEST", has_test ? mrb_true_value() : mrb_false_value());
-	mrb_define_global_const(mrb, "BTEST", has_btest ? mrb_true_value() : mrb_false_value());
+	mrb_define_global_const(mrb, "DEBUG", mrb_bool_value(shState->config().editor.debug));
```
Comment (Ancurio) [inline binding-mruby/binding-mruby.cpp]: Might want to set `TEST` here too according to `rgssVer` like you did in MRI, just for consistency.

**Comment by ReinUsesLisp** [comment]
git is kind of tricky sometimes. I used `git reset --soft ...` in the end, it took me some time to move away the filesystem commit (it was between commits).

**Comment by Ancurio** [comment]
Looks good to go; I'm just confused why your commit b022242 is still there. It looks like your master branch is based off of an out-of-date master (so why does github tell me "no conflicts when rebasing"?). I think I'll just pull this to see what happens, and correct my master afterwards manually.

Thanks!

**Comment by Ancurio** [comment]
Turns out, the overfluous commit just disappeared :) Nice

**Comment by ReinUsesLisp** [comment]
I manually re-commited filesystem's commit into b022242 after the "squash". Maybe that messed it up.

**Pull request merged**